File: src/java/samples/foundgine-supply-chain-advanced/src/main/java/com/foundgine/samples/supplychain/advanced/domain/Domain.java
package com.foundgine.samples.supplychain.advanced.domain;

import com.foundgine.providers.aot.*;
import java.math.BigDecimal;
import java.time.*;
import java.util.List;

public final class Domain {
  private Domain() {}
  public record CompanyId(int value) {} public record BusinessUnitId(int value) {}
  public record WarehouseId(int value) {} public record SupplierId(int value) {}
  public record SupplierSiteId(int value) {} public record CertificationId(int value) {}
  public record ProductId(int value) {} public record PurchaseOrderId(int value) {}
  public record PurchaseOrderLineId(int value) {} public record ShipmentId(int value) {}
  public record InventoryLotId(int value) {} public record CustomerOrderId(int value) {}
  public record CustomerOrderLineId(int value) {}

  @FoundgineEntity(name="Company") public record Company(int id,String name,@FoundgineSemanticDimension("tenant") String tenantId) {}
  @FoundgineEntity(name="BusinessUnit") public record BusinessUnit(int id,@FoundgineSemanticDimension("company") @FoundgineRelationship(target=Company.class,foreignKey="id",principalKey="companyId",name="company") int companyId,String name) {}
  @FoundgineEntity(name="Warehouse") public record Warehouse(@FoundgineRelationship(target=InventoryLot.class,foreignKey="warehouseId",principalKey="id",name="inventory") int id,@FoundgineSemanticDimension("businessUnit") @FoundgineRelationship(target=BusinessUnit.class,foreignKey="id",principalKey="businessUnitId",name="businessUnit") int businessUnitId,String name,@FoundgineSemanticDimension("tenant") String tenantId) {}
  @FoundgineEntity(name="Supplier") @FoundgineAlias(value="Vendor",weight=95) @FoundgineAlias(value="Seller",weight=90)
  public record Supplier(@FoundgineRelationship(target=SupplierCertification.class,foreignKey="supplierId",principalKey="id",name="certifications") int id,@FoundgineRelationship(target=ComplianceIncident.class,foreignKey="supplierId",principalKey="id",name="incidents") String name,@FoundgineSemanticDimension("country") String country,@FoundgineRelationship(target=PurchaseOrder.class,foreignKey="supplierId",principalKey="id",name="purchaseOrders") BigDecimal riskScore,@FoundgineSemanticDimension("tenant") String tenantId) {}
  public record SupplierSite(int id,int supplierId,@FoundgineSemanticDimension("country") String country,String name) {}
  @FoundgineEntity(name="SupplierCertification") public record SupplierCertification(int id,int supplierId,String type,LocalDate validFrom,LocalDate validTo) {}
  @FoundgineEntity(name="Product") public record Product(@FoundgineRelationship(target=ProductComponent.class,foreignKey="parentProductId",principalKey="id",name="components") int id,String sku,String name,@FoundgineSemanticDimension("category") String category,BigDecimal safetyStock,BigDecimal unitPrice) {}
  @FoundgineEntity(name="ProductComponent") public record ProductComponent(int parentProductId,@FoundgineRelationship(target=Product.class,foreignKey="id",principalKey="componentProductId",name="componentProduct") int componentProductId,BigDecimal quantityPerParent,LocalDate effectiveFrom,LocalDate effectiveTo,String revision,boolean phantom,BigDecimal yieldLossPercent,BigDecimal scrapFactor) {}
  public enum PurchaseOrderStatus { OPEN, PARTIALLY_RECEIVED, CANCELLED, CLOSED }
  @FoundgineEntity(name="PurchaseOrder") @FoundgineAlias(value="PO",weight=100) @FoundgineAlias(value="POs",weight=95) @FoundgineAlias(value="Buy",weight=90) @FoundgineAlias(value="Buys",weight=85)
  public record PurchaseOrder(@FoundgineRelationship(target=PurchaseOrderLine.class,foreignKey="purchaseOrderId",principalKey="id",name="lines") int id,@FoundgineRelationship(target=Supplier.class,foreignKey="id",principalKey="supplierId",name="supplier") int supplierId,@FoundgineRelationship(target=Shipment.class,foreignKey="purchaseOrderId",principalKey="id",name="shipments") int warehouseId,PurchaseOrderStatus status,LocalDate orderedOn,@FoundgineAlias(value="DueDate",weight=90) LocalDate expectedArrival) {}
  @FoundgineEntity(name="PurchaseOrderLine") public record PurchaseOrderLine(int id,@FoundgineRelationship(target=PurchaseOrder.class,foreignKey="id",principalKey="purchaseOrderId",name="purchaseOrder") int purchaseOrderId,int productId,BigDecimal quantity,BigDecimal unitPrice) {}
  public enum ShipmentStatus { PLANNED, IN_TRANSIT, DELAYED, PARTIALLY_RECEIVED, RECEIVED, CANCELLED }
  @FoundgineEntity(name="Shipment") public record Shipment(int id,int purchaseOrderId,LocalDate expectedArrival,LocalDate actualArrival,ShipmentStatus status,BigDecimal quantity) {}
  @FoundgineEntity(name="InventoryLot") public record InventoryLot(int id,@FoundgineRelationship(target=Warehouse.class,foreignKey="id",principalKey="warehouseId",name="warehouse") int warehouseId,int productId,BigDecimal onHand,BigDecimal reserved,BigDecimal quarantined,LocalDate receivedOn) {}
  public record Customer(int id,String name,String tenantId) {}
  @FoundgineEntity(name="Order") public record Order(@FoundgineRelationship(target=OrderItem.class,foreignKey="orderId",principalKey="id",name="items") int id,int customerId,String status,BigDecimal totalAmount,LocalDate placedOn) {}
  @FoundgineEntity(name="OrderItem") public record OrderItem(int id,int orderId,@FoundgineRelationship(target=Product.class,foreignKey="id",principalKey="productId",name="product") int productId,int quantity,BigDecimal unitPrice) {}
  public record OrderAllocation(int orderItemId,int lotId,int quantity) {}
  public record CancellationIdempotencyRecord(String key,String actor,int orderId,String requestFingerprint,BigDecimal restoredQuantity,LocalDate cancelledOn) {}
  public record IdempotencyRecord(String key,String actor,int customerId,int orderId,String requestFingerprint) {}
  public record CustomerOrder(int id,int businessUnitId,LocalDate placedOn,String status) {}
  @FoundgineEntity(name="CustomerOrderLine") public record CustomerOrderLine(int id,int customerOrderId,int productId,BigDecimal quantity) {}
  @FoundgineEntity(name="InventoryMovement") @FoundgineEvent(occurredAtField="OccurredAt") public record InventoryMovement(int id,int lotId,BigDecimal quantity,String reason,OffsetDateTime occurredAt) {}
  public record Allocation(int orderLineId,int lotId,BigDecimal quantity) {}
  public record ProductionOrder(int id,int productId,BigDecimal quantity,String status) {}
  public record ProductionMaterial(int id,int productionOrderId,int productId,BigDecimal quantity) {}
  @FoundgineEntity(name="ProductionOutput") @FoundgineEvent public record ProductionOutput(int id,int productionOrderId,int productId,BigDecimal quantity) {}
  @FoundgineEntity(name="ComplianceIncident") @FoundgineEvent(occurredAtField="OccurredOn") public record ComplianceIncident(int id,int supplierId,String severity,LocalDate occurredOn,String description) {}
}

File: src/csharp/samples/Foundgine.SupplyChain.Advanced/Semantic/Domain/Domain.cs
using Foundgine.Providers.Aot;
using Foundgine.Core.Semantic;

namespace Foundgine.SupplyChain.Advanced.Domain;

public readonly record struct CompanyId(int Value);

public readonly record struct BusinessUnitId(int Value);

public readonly record struct WarehouseId(int Value);

public readonly record struct SupplierId(int Value);

public readonly record struct SupplierSiteId(int Value);

public readonly record struct CertificationId(int Value);

public readonly record struct ProductId(int Value);

public readonly record struct PurchaseOrderId(int Value);

public readonly record struct PurchaseOrderLineId(int Value);

public readonly record struct ShipmentId(int Value);

public readonly record struct InventoryLotId(int Value);

public readonly record struct CustomerOrderId(int Value);

public readonly record struct CustomerOrderLineId(int Value);

/// <summary>Domain types are also the CLR source observed by the AOT metadata producer.</summary>
[FoundgineEntity("Company", StorageName = "companies", Id = 1000)]
public sealed record Company(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    CompanyId Id,
    string Name,
    [property: FoundgineSemanticDimension("tenant")]
    string TenantId);

[SemanticEntity]
[FoundgineEntity("BusinessUnit", StorageName = "business_units", Id = 1001)]
public sealed record BusinessUnit(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    BusinessUnitId Id,
    [property: FoundgineSemanticDimension("company"), FoundgineField(Index = true)]
    CompanyId CompanyId,
    string Name);

[SemanticEntity]
[FoundgineEntity("Warehouse", StorageName = "warehouses", Id = 1002)]
public sealed record Warehouse(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    WarehouseId Id,
    [property: FoundgineSemanticDimension("businessUnit"), FoundgineField(Index = true)]
    BusinessUnitId BusinessUnitId,
    string Name,
    [property: FoundgineSemanticDimension("tenant")]
    string TenantId)
{
    // Relationship ids are no longer assigned manually: when Id is omitted the
    // AOT generator derives a stable id from a hash of "Entity.Property", so
    // reordering members, merging modules, or adding new relationships
    // elsewhere never forces renumbering or risks a silent collision.
    [FoundgineRelationship(typeof(BusinessUnit), "BusinessUnitId", "Id", Name = "businessUnit")]
    public BusinessUnit BusinessUnit { get; init; } = null!;

    [FoundgineRelationship(typeof(InventoryLot), "WarehouseId", "Id", Name = "inventory")]
    public IReadOnlyList<InventoryLot> Inventory { get; init; } = [];
}

[SemanticEntity]
[FoundgineEntity("Supplier", StorageName = "suppliers", Id = 1003)]
[FoundgineAlias("Vendor", Weight = 95)]
[FoundgineAlias("Seller", Weight = 90)]
public sealed record Supplier(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    SupplierId Id,
    string Name,
    [property: FoundgineSemanticDimension("country"), FoundgineField(Index = true),
               FoundgineAlias("State", Weight = 85)]
    string Country,
    decimal RiskScore,
    [property: FoundgineSemanticDimension("tenant")]
    string TenantId)
{
    [FoundgineRelationship(typeof(SupplierCertification), "SupplierId", "Id", Name = "certifications")]
    public IReadOnlyList<SupplierCertification> Certifications { get; init; } = [];

    [FoundgineRelationship(typeof(ComplianceIncident), "SupplierId", "Id", Name = "incidents")]
    public IReadOnlyList<ComplianceIncident> Incidents { get; init; } = [];

    // Reverse semantic relationship (recommendation 3): lets a query start
    // from a supplier and ask "which purchase orders depend on this
    // supplier?" without the caller having to know PurchaseOrder owns the
    // foreign key.
    [FoundgineRelationship(typeof(PurchaseOrder), "SupplierId", "Id", Name = "purchaseOrders")]
    public IReadOnlyList<PurchaseOrder> PurchaseOrders { get; init; } = [];
}

[FoundgineEntity("SupplierSite", StorageName = "supplier_sites", Id = 1004)]
public sealed record SupplierSite(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    SupplierSiteId Id,
    [property: FoundgineField(Index = true)]
    SupplierId SupplierId,
    [property: FoundgineSemanticDimension("country")]
    string Country,
    string Name);

[SemanticEntity]
[FoundgineEntity("SupplierCertification", StorageName = "supplier_certifications", Id = 1005)]
public sealed record SupplierCertification(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    CertificationId Id,
    [property: FoundgineField(Index = true)]
    SupplierId SupplierId,
    string Type,
    DateOnly ValidFrom,
    DateOnly ValidTo);

[SemanticEntity]
[FoundgineEntity("Product", StorageName = "products", Id = 1006)]
public sealed record Product(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    ProductId Id,
    string Sku,
    string Name,
    [property: FoundgineSemanticDimension("category"), FoundgineField(Index = true)]
    string Category,
    decimal SafetyStock)
{
    [FoundgineRelationship(typeof(ProductComponent), "ParentProductId", "Id", Name = "components")]
    public IReadOnlyList<ProductComponent> Components { get; init; } = [];

    [FoundgineRelationship(typeof(PurchaseOrderLine), "ProductId", "Id", Name = "purchaseOrderLines")]
    public IReadOnlyList<PurchaseOrderLine> PurchaseOrderLines { get; init; } = [];

    // Reverse semantic relationship (recommendation 3): "where is this used?"
    // - every BOM line that consumes this product as a component.
    [FoundgineRelationship(typeof(ProductComponent), "ComponentProductId", "Id", Name = "usedInComponents")]
    public IReadOnlyList<ProductComponent> UsedInComponents { get; init; } = [];
}

[SemanticEntity]
[FoundgineEntity("ProductComponent", StorageName = "product_components", Id = 1007)]
public sealed record ProductComponent(
    [property: FoundgineField("ParentProductId", Id = 1, IsPrimaryKey = true)]
    ProductId ParentProductId,
    [property: FoundgineField(Index = true)]
    ProductId ComponentProductId,
    decimal QuantityPerParent,
    // BOM metadata (recommendation 4): even where unused today, these are the
    // fields a manufacturing/planning engine eventually needs from a BOM line.
    DateOnly? EffectiveFrom = null,
    DateOnly? EffectiveTo = null,
    string Revision = "1",
    bool IsPhantom = false,
    decimal YieldLossPercent = 0m,
    decimal ScrapFactor = 0m)
{
    [FoundgineRelationship(typeof(Product), "ComponentProductId", "Id", Name = "componentProduct")]
    public Product ComponentProduct { get; init; } = null!;
}

public enum PurchaseOrderStatus
{
    Open,
    PartiallyReceived,
    Cancelled,
    Closed
}

[SemanticEntity]
[FoundgineEntity("PurchaseOrder", StorageName = "purchase_orders", Id = 1008)]
[FoundgineAlias("PO", Weight = 100)]
[FoundgineAlias("POs", Weight = 95)]
[FoundgineAlias("Buy", Weight = 90)]
[FoundgineAlias("Buys", Weight = 85)]
public sealed record PurchaseOrder(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    PurchaseOrderId Id,
    [property: FoundgineField(Index = true)]
    SupplierId SupplierId,
    [property: FoundgineField(Index = true)]
    WarehouseId WarehouseId,
    PurchaseOrderStatus Status,
    DateOnly OrderedOn,
    [property: FoundgineAlias("DueDate", Weight = 90)]
    DateOnly ExpectedArrival)
{
    [FoundgineRelationship(typeof(PurchaseOrderLine), "PurchaseOrderId", "Id", Name = "lines")]
    public IReadOnlyList<PurchaseOrderLine> Lines { get; init; } = [];

    [FoundgineRelationship(typeof(Shipment), "PurchaseOrderId", "Id", Name = "shipments")]
    public IReadOnlyList<Shipment> Shipments { get; init; } = [];

    [FoundgineRelationship(typeof(Supplier), "SupplierId", "Id", Name = "supplier")]
    [FoundgineAlias("vendor", Weight = 85)]
    public Supplier Supplier { get; init; } = null!;
}

[SemanticEntity]
[FoundgineEntity("PurchaseOrderLine", StorageName = "purchase_order_lines", Id = 1009)]
public sealed record PurchaseOrderLine(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    PurchaseOrderLineId Id,
    PurchaseOrderId PurchaseOrderId,
    [property: FoundgineField(Index = true)]
    ProductId ProductId,
    decimal Quantity,
    decimal UnitPrice)
{
    [FoundgineRelationship(typeof(PurchaseOrder), "PurchaseOrderId", "Id", Name = "purchaseOrder")]
    public PurchaseOrder PurchaseOrder { get; init; } = null!;
}

public enum ShipmentStatus
{
    Planned,
    InTransit,
    Delayed,
    PartiallyReceived,
    Received,
    Cancelled
}

[SemanticEntity]
[FoundgineEntity("Shipment", StorageName = "shipments", Id = 1010)]
public sealed record Shipment(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    ShipmentId Id,
    PurchaseOrderId PurchaseOrderId,
    DateOnly ExpectedArrival,
    DateOnly? ActualArrival,
    ShipmentStatus Status,
    decimal Quantity);

[SemanticEntity]
[FoundgineEntity("InventoryLot", StorageName = "inventory", Id = 1011)]
public sealed record InventoryLot(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    InventoryLotId Id,
    WarehouseId WarehouseId,
    [property: FoundgineField(Index = true)]
    ProductId ProductId,
    decimal OnHand,
    decimal Reserved,
    decimal Quarantined,
    DateOnly ReceivedOn)
{
    [FoundgineRelationship(typeof(Warehouse), "WarehouseId", "Id", Name = "warehouse")]
    public Warehouse Warehouse { get; init; } = null!;
}

[SemanticEntity]
[FoundgineEntity("CustomerOrder", StorageName = "customer_orders", Id = 1012)]
public sealed record CustomerOrder(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    CustomerOrderId Id,
    BusinessUnitId BusinessUnitId,
    DateOnly PlacedOn,
    string Status)
{
    [FoundgineRelationship(typeof(CustomerOrderLine), "CustomerOrderId", "Id", Name = "lines")]
    public IReadOnlyList<CustomerOrderLine> Lines { get; init; } = [];
}

[SemanticEntity]
[FoundgineEntity("CustomerOrderLine", StorageName = "customer_order_lines", Id = 1013)]
public sealed record CustomerOrderLine(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    CustomerOrderLineId Id,
    CustomerOrderId CustomerOrderId,
    ProductId ProductId,
    decimal Quantity);

// Event entities (recommendation 5): each of these records something that
// happened at a point in time and is never mutated afterward, as opposed to
// the state entities above (Product, Supplier, InventoryLot, ...) whose rows
// are updated in place. [FoundgineEvent] marks that distinction in metadata
// and, where given a field name, points at the column a temporal/"as of"
// query should key on.
[FoundgineEntity("InventoryMovement", StorageName = "inventory_movements", Id = 1015)]
[FoundgineEvent("OccurredAt")]
public sealed record InventoryMovement(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    int Id,
    [property: FoundgineField(Index = true)]
    InventoryLotId LotId,
    decimal Quantity,
    string Reason,
    DateTimeOffset OccurredAt);

public sealed record Allocation(CustomerOrderLineId OrderLineId, InventoryLotId LotId, decimal Quantity);

public sealed record ProductionOrder(int Id, ProductId ProductId, decimal Quantity, string Status);

public sealed record ProductionMaterial(int Id, int ProductionOrderId, ProductId ProductId, decimal Quantity);

[FoundgineEntity("ProductionOutput", StorageName = "production_outputs", Id = 1016)]
[FoundgineEvent]
public sealed record ProductionOutput(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    int Id,
    int ProductionOrderId,
    ProductId ProductId,
    decimal Quantity);

[SemanticEntity]
[FoundgineEntity("ComplianceIncident", StorageName = "compliance_incidents", Id = 1014)]
[FoundgineEvent("OccurredOn")]
public sealed record ComplianceIncident(
    [property: FoundgineField("Id", Id = 1, IsPrimaryKey = true)]
    int Id,
    [property: FoundgineField(Index = true)]
    SupplierId SupplierId,
    string Severity,
    DateOnly OccurredOn,
    string Description);